osh/cmd_eval.py

```diff
@@ -1896,7 +1896,12 @@ def ExecuteAndCatch(self, node, cmd_flags=0):
         """
         if cmd_flags & Optimize:
             node = self._RemoveSubshells(node)
-            self._NoForkLast(node)  # turn the last ones into exec
+            # Using _NoForkLast() hides exit code for ERR trap, temporarily comment it
```

Comment:
Oh NoForkLast actually breaks  a bunch of other stuff

I think we need `shopt --unset noforklast` or something like that

https://www.oilshell.org/blog/2024/03/release-0.21.0.html#starship-prompt-bug-pipelines-crash-after-using-it


I would leave it off this change

The way to work around it is to add the `;true` -- that's fine

Let's tackle this separately, since it's a crosscutting issue

I definitely welcome help on it though!!

I think this came up in one of @melvinw PR's too.  It is a problematic thing that all shells implement, slightly differently

Reply:
For the moment I was not able to find a way to get the exit code of the last command, unless I remove this option.
Once unset, all issues magically disappear, except one regarding pipeline I need to work on.

But if course, if there is another way to get the exit code and maybe call the err trap, then we should be able to remove it.

Reply:
I think a summary of the state of this PR can be interesting, although I understand there are still some open topics you've raised and that are not closed yet.

Analysing the CI tests results, there's a very few issues remaining, most of them being not real, or related to other fixes we can work on.

**test/runtime-errors.sh test-errexit-multiple-processes**

Is due to the errors display, this can be fixed in another specific branch, I have an idea on how to fix it to have only one single error displayed, because I guess the idea is to use verbose_errexit to show only one single error, and not many of them. Currently 3 errors are displayed instead of only 1 :

```
  ls | false | wc -l
       ^~~~~
[ -c flag ]:1: errexit PID 354396: command.Simple failed with status 1
  ls | false | wc -l
  ^~
[ -c flag ]:1: errexit PID 354395: command.Simple failed with status 141
0
  ls | false | wc -l
       ^~~~~
[ -c flag ]:1: errexit PID 354392: command.Pipeline failed with status 1
```

The first because it's the original error (and the first one), the second should be hidden as it is a SIGPIPE, so we might know we're in sigpipe to prevent, or maybe always prevent sigpipe errors, don't know. The third because the result of the pipeline is in error.

The idea is while parsing the AST, to identify which parts are allowed to display errexit errors and which should not, and once the first error is displayed,, just stop displaying anything. This is the idea... Maybe add a field to the mem object, or in mutable_opts?

**test/spec-py.sh ysh-all-serial**

The NoForkLast change impacted the way the xtrace is displaying things, this is just the test to be updated to reflect the change.

**test/spec.sh builtin-trap-bash -r 19 -v**

This one is expected failing, but I think I got the reason why now : the "dbg 3" is printed on stdout, which is eaten by the pipeline of `false | false | false`. To verify this, just do `trap 'echo dbg $LINENO > /dev/stderr' DEBUG` and here comes the 3 `dbg 3`.

I guess bash and other shells are doing tricks to make the debug to go the the stdout of the pipeline and not the sub commands. Probably this another hidden issue we should also fix, as it may have other side effects not yet identified.
A dedicated branch for this one too?

**test/stateful.sh job-control-all -r 10**

This one is the only remaining issue I think there is. I didn't started yet, as I was looking to other subjects today.